### PR TITLE
e2e tests for Federated Secrets

### PR DESCRIPTION
**Pull request by kshafiee**
e2e test to confirm that underlying secrets are always created correctly in underlying clusters

This is to address Issue #29337

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31655)

### CODE REVIEW

**Comment by kshafiee** [comment]
@nikhiljindal @quinton-hoole @jianhuiz @deepak-vij

**Comment by kshafiee** [comment]
ok to test

**Comment by quinton-hoole** [comment]
ok to test

**Comment by quinton-hoole** [comment]
@k8s-bot federation gce e2e test this

**Comment by quinton-hoole** [comment]
You can ignore the "Jenkins Federation GCE e2e" error - it's unrelated to this PR.  I've filed an issue to have it fixed. https://github.com/kubernetes/test-infra/issues/470

**Comment by kshafiee** [comment]
Thanks @quinton-hoole , sure!

File: test/e2e/federated-secret.go [hunk 1]
```diff
@@ -35,52 +36,111 @@ const (
...
-				Expect(err).NotTo(HaveOccurred())
+			deleteAllTestSecrets(
+				f.FederationClientset_1_4.Core().Secrets(nsName).List,
+				f.FederationClientset_1_4.Core().Secrets(nsName).Delete,
+				false)
+			for _, clientset := range clusterClientSet {
+				deleteAllTestSecrets(
+					clientset.Core().Secrets(nsName).List,
+					clientset.Core().Secrets(nsName).Delete,
+					false)
```
Comment (quinton-hoole) [inline test/e2e/federated-secret.go]: Ideally you need a call to unregisterClusters at the end of ForEach().  See 

https://github.com/kubernetes/kubernetes/blob/8c110675406ff822e7245563f9911fb173a0015e/test/e2e/federated-ingress.go#L101

and 

https://github.com/kubernetes/kubernetes/blob/4edc98260f26b2047ae9fcfa33cca1554dd1a3b2/test/e2e/federation-util.go#L163

File: test/e2e/federated-secret.go [hunk 2]
```diff
@@ -35,52 +36,111 @@ const (
...
 			}
 		})
 
 		It("should be created and deleted successfully", func() {
 			framework.SkipUnlessFederated(f.Client)
 
 			nsName := f.FederationNamespace.Name
-			secret := createSecretOrFail(f.FederationClientset_1_4, nsName)
+
```
Comment (quinton-hoole) [inline test/e2e/federated-secret.go]: Why this change?

File: test/e2e/federated-secret.go [hunk 3]
```diff
@@ -35,52 +36,111 @@ const (
...
+				return true, nil
+			})
+			framework.ExpectNoError(err, "Not all secrets created")
+
 			By(fmt.Sprintf("Creation of secret %q in namespace %q succeeded.  Deleting secret.", secret.Name, nsName))
 			// Cleanup
-			err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-			framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, secret.Namespace)
-			By(fmt.Sprintf("Deletion of secret %q in namespace %q succeeded.", secret.Name, nsName))
+			deleteAllTestSecrets(
```
Comment (quinton-hoole) [inline test/e2e/federated-secret.go]: You already have this in AfterEach(), so you shouldn't need it here?

File: test/e2e/federated-secret.go [hunk 4]
```diff
@@ -35,52 +36,111 @@ const (
...
+				ObjectMeta: v1.ObjectMeta{
+					Name: FederatedSecretName,
+				},
+			}
+
+			By(fmt.Sprintf("Creating secret %q in namespace %q", secret.Name, nsName))
+			_, err := f.FederationClientset_1_4.Core().Secrets(nsName).Create(secret)
+			framework.ExpectNoError(err, "Failed to create secret %s", secret.Name)
+
+			// Check subclusters if the secret was created there
```
Comment (quinton-hoole) [inline test/e2e/federated-secret.go]: nit: I would prefer this in a separate function, so that it can be more easily used in future.

**Comment by quinton-hoole** [comment]
Thanks @kshafiee .  A few small changes, and then LGTM.  We will also need an update test (which should check that the update is visible in the federation, and also eventually in the underlying clusters).  Please file an issue for that, and if possible, start work on it.

**Comment by kshafiee** [comment]
ok to test

File: test/e2e/federated-secret.go (hunk 1 above)
Comment (kshafiee) [inline test/e2e/federated-secret.go]: I am using function 'buildClustersOrFail_14' in 'federation-util-14.go' that returns clusters of type '[]*federation_api.Cluster'
In the federated-ingress.go example Jesse is using function 'registerClusters' in 'federation-util.go' that takes clusters of type 'map[string]*cluster'

The reason I used 'federation-util-14.go' was I saw federated-namespace.go as an example. Please let me know which one I should use. Thx!

Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: Yes unfortunately we have ended up having those 2 util files that do the same thing in slightly different ways.
We want to consolidate them.

If you update this code to use federation-util.go, then feel free to make the same change to the namespace test. Ref: https://github.com/kubernetes/kubernetes/pull/30443#issuecomment-239492984.

cc @mwielgus

**Comment by quinton-hoole** [comment]
@kshafiee Is this ready for re-review and merge?  Let me know and I'll do it.

**Comment by kshafiee** [comment]
Thanks @quinton-hoole for your comments!

- I updated the code to use the functions in federation-util.go
- I added an update test

It is good for re-review now!

File: test/e2e/federated-secret.go [hunk 5]
```diff
@@ -18,69 +18,161 @@ package e2e
...
+}
+
+func updateSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (newSecret *v1.Secret, err error) {
+	secret := &v1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name: UpdatedFederatedSecretName,
+		},
+	}
+
+	for MaxRetriesOnFederatedApiserver := 0; MaxRetriesOnFederatedApiserver < 3; MaxRetriesOnFederatedApiserver++ {
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: MaxRetries could be a constant at the top alongside FederatedSecretTimeout and use a variable retryCount here?

File: test/e2e/federated-secret.go [hunk 6]
```diff
@@ -18,69 +18,161 @@ package e2e
...
 	})
 })
 
-func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
-	if clientset == nil || len(namespace) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteSecretOrFail: clientset: %v, namespace: %v", clientset, namespace))
-	}
-	By(fmt.Sprintf("Creating federated secret %q in namespace %q", FederatedSecretName, namespace))
-
+func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (*v1.Secret, error) {
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: You seem to be ignoring the returned err everywhere assuming the function will fail if there is an error.
No need to return an error in that case?

Same for the updateSecret fn

File: test/e2e/federated-secret.go [hunk 7]
```diff
@@ -18,69 +18,161 @@ package e2e
...
+			framework.SkipUnlessFederated(f.Client)
+			if federationName = os.Getenv("FEDERATION_NAME"); federationName == "" {
+				federationName = DefaultFederationName
 			}
+			clusters = map[string]*cluster{}
+			_ = registerClusters(clusters, UserAgentName, federationName, f)
+		})
+
+		AfterEach(func() {
+			unregisterClusters(clusters, f)
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: Add framework.SkipUnlessFederated(f.Client)

File: test/e2e/federated-secret.go [hunk 8]
```diff
@@ -18,69 +18,161 @@ package e2e
...
-			for _, secret := range secretList.Items {
-				err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-				Expect(err).NotTo(HaveOccurred())
+		BeforeEach(func() {
+			framework.SkipUnlessFederated(f.Client)
+			if federationName = os.Getenv("FEDERATION_NAME"); federationName == "" {
+				federationName = DefaultFederationName
 			}
+			clusters = map[string]*cluster{}
+			_ = registerClusters(clusters, UserAgentName, federationName, f)
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: registerClusters is not using the federationName. remove it?

**Comment by kshafiee** [comment]
Thanks @nikhiljindal ! Let me know if your comments are addressed.

File: test/e2e/federated-secret.go [hunk 9]
```diff
@@ -18,69 +18,157 @@ package e2e
...
+
+func updateSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) (newSecret *v1.Secret) {
+	var err error
+	secret := &v1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name: UpdatedFederatedSecretName,
+		},
+	}
+
+	for RetryCount := 0; RetryCount < MaxRetries; RetryCount++ {
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: RetryCount -> retryCount

File: test/e2e/federated-secret.go [hunk 10]
```diff
@@ -18,69 +18,157 @@ package e2e
...
+			// wait for secret shards being created
+			waitForSecretShardsOrFail(nsName, secret, clusters)
+			secret = updateSecretOrFail(f.FederationClientset_1_4, nsName)
+			waitForSecretShardsUpdatedOrFail(nsName, secret, clusters)
+		})
 	})
 })
 
 func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
-	if clientset == nil || len(namespace) == 0 {
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: Why delete this?
Seems useful to keep.

**Comment by nikhiljindal** [comment]
Thanks @kshafiee 
2 minor comments, mostly LGTM

**Comment by kshafiee** [comment]
Thx @nikhiljindal comments addressed!

**Comment by nikhiljindal** [comment]
LGTM, thanks

**Comment by kshafiee** [comment]
ok to test

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this issue: #IGNORE

File: test/e2e/federated-secret.go [hunk 11]
```diff
@@ -18,69 +18,166 @@ package e2e
...
+			Name: UpdatedFederatedSecretName,
+		},
+	}
+
+	for retryCount := 0; retryCount < MaxRetries; retryCount++ {
+		_, err = clientset.Core().Secrets(namespace).Get(FederatedSecretName)
+		if err != nil {
+			framework.Failf("failed to get secret %q: %v", FederatedSecretName, err)
+		}
+		newSecret, err = clientset.Core().Secrets(namespace).Update(secret)
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: This command is failing with:
```
failed to update secret "federated-secret": secrets "updated-federated-secret" not found
```

We cannot change the name of a resource while updating it.

cc @kshafiee
